jax/interpreters/pxla.py

```diff
@@ -71,77 +72,20 @@
 Index = Union[int, slice, Tuple[Union[int, slice], ...]]
 
 
-class NoSharding:
-
-  def __eq__(self, other):
-    return isinstance(other, NoSharding)
-
-  def __repr__(self):
-    return "NoSharding()"
-
+# See the C++ code for comments.
+NoSharding = pmap_lib.NoSharding
+Chunked = pmap_lib.Chunked
+Unstacked = pmap_lib.Unstacked
 
 _UNSHARDED_INSTANCE = NoSharding()
 
-# mypy is very unhappy about taggedtuple
-if TYPE_CHECKING:
-  class Unstacked(NamedTuple):
-    size: int
-else:
-  Unstacked = taggedtuple('Unstacked', ('size',))
-
-class Chunked:
-  chunks: Tuple[int, ...]
-
-  def __init__(self, chunks: Union[int, Tuple[int, ...]]):
-    if not isinstance(chunks, tuple):
-      chunks = (chunks,)
-    object.__setattr__(self, 'chunks', chunks)
-
-  def __setattr__(self, name, value):
-    raise RuntimeError("Chunked is immutable")
-
-  def __delattr__(self, name):
-    raise RuntimeError("Chunked is immutable")
-
-  def __hash__(self):
-    return hash(self.chunks)
-
-  def __eq__(self, other):
-    return type(other) is Chunked and self.chunks == other.chunks
-
-  def __repr__(self):
-    return f'Chunked({self.chunks})'
-
-"""
-Represents all the ways we can shard a dimension.
-- `None` means no sharding;
-- `Chunked` means that the dimension is split into the specified number of chunks,
-  but the split dimension itself is preserved inside the map;
-- `Unstacked` means that the dimension is split into chunks of size 1, and doesn't
-  appear inside the map.
-"""
-AvalDimSharding = Union[Unstacked, Chunked, NoSharding]
-
-# mypy is very unhappy about taggedtuple
-if TYPE_CHECKING:
-  class ShardedAxis(NamedTuple):
-    axis: int
-  class Replicated(NamedTuple):
-    replicas: int
-else:
-  ShardedAxis = taggedtuple('ShardedAxis', ('axis',))
-  Replicated = taggedtuple('Replicated', ('replicas',))
+ShardedAxis = pmap_lib.ShardedAxis
+Replicated = pmap_lib.Replicated
 
-"""
-Assigns sharded axes to mesh dimensions.
+# mypy cannot deal with the C++ types.
+AvalDimSharding = Union[Unstacked, Chunked, NoSharding]  # type: ignore
+MeshDimAssignment = Union[ShardedAxis, Replicated]  # type: ignore
```

Comment:
Instead of `# type: ignore`, you could use the `if TYPE_CHECKING` trick I used. That would be a bit better because it would at least give us better coverage.

Reply:
Done.